### ISSUE TITLE
Nose is loaded in production

### ISSUE THREAD

**Issue by cclamb** [Theano/Theano#3096]
Hey everybody, nose is loading in production, and it breaks interaction with Theano and ipdb in ipython notebooks (see https://github.com/gotcha/ipdb/issues/52).

When I load theano into ipython or the python REBL, I get this (or equivalent, this is in the standard python REPL):

```
Python 2.7.10 (default, Jun 10 2015, 19:42:47) 
[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> msg = 'got nose!' if 'nose' in sys.modules.keys() else 'no nose.'
>>> print msg
no nose.
>>> import theano
>>> import sys
>>> msg = 'got nose!' if 'nose' in sys.modules.keys() else 'no nose.'
>>> print msg
got nose!
>>> 
```

I've hacked around this, but you might want to consider removing nose references from your distributed builds.

**Comment by nouiz** [comment]
Do you want to work on it? We are pretty busy and won't work on this shortly.

If you can't work on it, can you paste the diff of your work around to be sure to guide the fix of that problem?

thanks for the report.

**Comment by cclamb** [comment]
Sure, I'll see what I can do. The workaround is in the cited ticket over on the ipdb side.

**Comment by lamblin** [comment]
The unit tests are part of the distributed builds, and they depend on `nose` for things like `SkipTest` for instance.
However, it is true that the tests should not be imported when simply importing `theano`.
Likely culprits are:
- `theano.tests.unittest_tools`, which may get imported from outside
- `theano.sandbox.cuda`, which imports `test_driver` from the `tests` subdirectory. `test_driver` should probably be moved somewhere else.

**Comment by mohammadpz** [comment]
@abergeron , This happens at lines 92 to 100 of ```theano/__init__.py```:
```python
# This need to be before the init of GPU, as it add config variable
# needed during that phase.
import theano.tests
if hasattr(theano.tests, "TheanoNoseTester"):
    test = theano.tests.TheanoNoseTester().test
else:
    def test():
        raise ImportError("The nose module is not installed."
                          " It is needed for Theano tests.")
```
When we import ```import theano.tests```, it imports nose too.

Any idea on how to avoid importing tests or importing then whenever needed?

**Comment by abergeron** [comment]
Maybe create a function `test()` which does the above imports/tests?

**Comment by abergeron** [comment]
Fixed by #3389
